File: algorithms/hard/maximum_performance_of_a_team.ex
defmodule Solution do
  @spec max_performance(n :: integer, speed :: [integer], efficiency :: [integer], k :: integer) ::
          integer
  def max_performance(n, speed, efficiency, k) do
    Enum.zip(efficiency, speed)
    |> Enum.sort(&>=/2)
    |> Enum.reduce({0, 0, :gb_sets.new()}, fn {e, s}, {ans, total_speed, pq} ->
      {total_speed, pq} =
        if :gb_sets.size(pq) == k do
          {{min_speed, _}, pq} = :gb_sets.take_smallest(pq)
          {total_speed - min_speed, pq}
        else
          {total_speed, pq}
        end

      total_speed = total_speed + s
      {max(ans, total_speed * e), total_speed, :gb_sets.add({s, :random.uniform()}, pq)}
    end)
    |> elem(0)
    |> rem(1_000_000_007)
  end
end

File: algorithms/hard/maximum_performance_of_a_team.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;

impl Solution {
    pub fn max_performance(n: i32, speed: Vec<i32>, efficiency: Vec<i32>, k: i32) -> i32 {
        let mut engineers = Vec::new();
        for i in 0..speed.len() {
            engineers.push((efficiency[i] as i64, speed[i] as i64));
        }
        engineers.sort_by(|a, b| b.cmp(a));

        let mut ans = 0;
        let mut total_speed = 0;
        let mut pq: BinaryHeap<Reverse<_>> = BinaryHeap::new();
        for (e, s) in engineers {
            if pq.len() == (k as usize) {
                total_speed -= pq.pop().unwrap().0;
            }
            pq.push(Reverse(s));
            total_speed += s;
            ans = ans.max(total_speed * e);
        }
        (ans % 1000000007) as i32
    }
}
